src/cloudflare/internal/pipeline-transform.ts

```diff
@@ -0,0 +1,151 @@
+  shard: string; // assigned shard
+  ts: number; // creation timestamp of the batch
+
+  format: Format;
+  size: {
+    bytes: number;
+    rows: number;
+  };
+  data: unknown;
+};
+
+type JsonStream = Batch & {
+  format: Format.JSON_STREAM;
+  data: ReadableStream<Uint8Array>;
+};
+
+enum Format {
+  JSON_STREAM = 'json_stream', // jsonl
+}
+
+export class PipelineTransformImpl extends entrypoints.WorkerEntrypoint {
+  #batch?: Batch;
+  #initalized: boolean = false;
+
+  // stub overriden on the sub class
+  // eslint-disable-next-line @typescript-eslint/require-await
+  public async transformJson(_data: object[]): Promise<object[]> {
+    throw new Error('should be implemented by parent');
+  }
+
+  // called by the dispatcher which then calls the subclass methods
+  // @ts-expect-error thinks ping is never used
+  private async _ping(): Promise<void> {
+    // making sure the function was overriden by an implementing subclass
+    if (this.transformJson !== PipelineTransformImpl.prototype.transformJson) {
+      return await Promise.resolve(); // eslint
+    } else {
+      throw new Error(
+        'the transformJson method must be overridden by the PipelineTransform subclass'
+      );
+    }
+  }
+
+  // called by the dispatcher which then calls the subclass methods
+  // the reason this is typescript private and not javascript private is that this must be
+  // able to be called by the dispatcher but should not be called by the class implementer
+  // @ts-expect-error _transform is called by rpc
+  private async _transform(batch: Batch): Promise<JsonStream> {
+    if (this.#initalized) {
+      throw new Error('pipeline entrypoint has already been initialized');
+    }
+
+    this.#batch = batch;
+    this.#initalized = true;
+
+    switch (this.#batch.format) {
+      case Format.JSON_STREAM:
+        const data = await this.#readJsonStream();
+        const transformed = await this.transformJson(data);
+        return this.#sendJson(transformed);
+      default:
+        throw new Error('unsupported batch format');
+    }
+  }
+
+  async #readJsonStream(): Promise<object[]> {
+    if (this.#batch!.format !== Format.JSON_STREAM) {
+      throw new Error(`expected JSON_STREAM not ${this.#batch!.format}`);
+    }
+
+    const batch = this.#batch!.data as ReadableStream<Uint8Array>;
+    const decoder = batch.pipeThrough(new TextDecoderStream());
+
+    const data: object[] = [];
+    for await (const line of readLines(decoder)) {
+      data.push(JSON.parse(line) as object);
+    }
+
+    return data;
+  }
+
+  #sendJson(data: object[]): JsonStream {
+    if (!(data instanceof Array)) {
+      throw new Error('transformJson must return an array of objects');
+    }
+
+    let written = 0;
+    const encoder = new TextEncoder();
+    const readable = new ReadableStream<Uint8Array>({
+      start(controller): void {
+        for (const obj of data) {
+          const encoded = encoder.encode(`${JSON.stringify(obj)}\n`);
```

Comment:
In terms of validating the response from the customer's code, it's [also possible for them to return values that will throw when passed to `JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#exceptions), and we'd *probably* benefit from handling them more gracefully than just an uncaught exception?

Although then again in the `instanceof Array` check above all we do is throw an exception, so actually maybe this is already fine and we just need the calling code to detect this case correctly?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#exceptions